### PR TITLE
Fix continuous replication reachablity issue

### PR DESCRIPTION
**Pull request by pasin**
## Changes:

- Always try to bring the replication online first time regardless of the replicator mode

- When the replication is in the continuous mode and the protocol is not CBL_URLProtocol, start the reachability. The reachability is now responsible for 2 things when the host is reachable (Ignore not-reachable status, which could be false negative):

 (1) It checks if the host is reachable on an allowed network (based on the kCBLReplicatorOption_Network option value). If the host is reachable in a forbidden network, it will bring the replicator offline.

 (2) It triggers the replicator retry logic if the replication is already online. If the replicator is not online, it will bring the replicator online.

- When `ChangeTracker` or `CBLRemoteRequest` gets a `reachability` error, it will report the error immediately without retrying. Once the error gets propagated to the CBL_Replicator, the CBL_Replicator will bring the replicator offline manually and then it will try to bring it online after 5 mins.

- Expand CBLIsOfflineError to cover more `reachability` error codes.

## Note:
The reachability is still using to restrict the network access based on value set for kCBLReplicatorOption_Network option. This is not perfect as the reachability sometimes reports a wrong status. In another word, I'm ignoring this issue for now. To fix this we can set the NSMutableURLRequest's allowsCellularAccess property and CFStream's kCFStreamPropertyConnectionIsCellular property instead of using the reachability. However the solution is not completed as we are using CocoaAsycSocket for the WebSocket version of the ChangeTracker and we couldn't find a way to restrict the cellular network access.

#536

### CODE REVIEW

**Comment by pasin** [comment]
Just notice that CBL_Puller also calls the goOffline() method by itself when its change tracker receives a reachability error (offline error). I need to fix this and propagate the error to CBL_Replicator so it can kick off a retry cycle. Taking a PR back to fix this.

**Comment by pasin** [comment]
Resubmitted the PR again.

File: Source/CBL_Replicator.m [hunk 1]
```diff
@@ -465,8 +476,8 @@ - (void) reachabilityChanged: (CBLReachability*)host {
 
     BOOL reachable = host.reachable && !_suspended;
     if (reachable) {
-    // Parse "network" option. Could be nil or "WiFi" or "!Wifi" or "cell" or "!cell".
-    NSString* network = [$castIf(NSString, _options[kCBLReplicatorOption_Network])
+        // Parse "network" option. Could be nil or "WiFi" or "!Wifi" or "cell" or "!cell".
+        NSString* network = [$castIf(NSString, _options[kCBLReplicatorOption_Network])
```
Comment (snej) [inline Source/CBL_Replicator.m]: Oops, looks like there are tab characters here.

Comment (pasin) [inline Source/CBL_Replicator.m]: When I look at that in my xcode, it's spaces.

**Comment by pasin** [comment]
@snej , I tested more with my test app and fixed an issue that the pusher kept itself in an offline mode when coming back from the background mode while no internet connection and then having internet connection after. Basically when the app coming from the background mode, we need to try to goOnline right way as well. I revised the methods and variable names. If you have time, please review the PR.

**Comment by pasin** [comment]
Taking the PR back. The solution alters the current behavior of the replication. A simpler solution might be just trying to connect to the server when the remote host is not reachable reported by the SCNetworkReachability.

**Pull request closed without merging**